core/services/pipeline/runner.go

```diff
@@ -312,14 +387,69 @@ func logTaskRunToPrometheus(trr TaskRunResult, spec Spec) {
 func (r *runner) ExecuteAndInsertFinishedRun(ctx context.Context, spec Spec, vars Vars, l logger.Logger, saveSuccessfulTaskRuns bool) (runID int64, finalResult FinalResult, err error) {
 	run, trrs, err := r.ExecuteRun(ctx, spec, vars, l)
 	if err != nil {
-		return run.ID, finalResult, errors.Wrapf(err, "error executing run for spec ID %v", spec.ID)
+		return 0, finalResult, errors.Wrapf(err, "error executing run for spec ID %v", spec.ID)
 	}
-	finalResult = trrs.FinalResult()
-	runID, err = r.orm.InsertFinishedRun(r.orm.DB(), run, trrs, saveSuccessfulTaskRuns)
-	if err != nil {
-		return runID, finalResult, errors.Wrapf(err, "error inserting finished results for spec ID %v", spec.ID)
+
+	if run.Async {
```

Comment:
Can we make this more readable by splitting out the body of the if into functions?